### PR TITLE
Added smart attribute keeping when changing label

### PR DESCRIPTION
**Pull request by timurlenk07**
### Motivation and context

Implements #2755 and #3294. Basically, when changing label class, attributes are retained **IF and only** there is an attribute with the same name and type, and the old value is still valid in the new configuration.

### How has this been tested?

Tested manually on rectangle labels with various attributes.

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by timurlenk07** [comment]
As I work with images, I did not know whether attribute mutability should be considered in the attribute keeping decision. Someone better informed could tell me, what would be the correct way to handle that?

File: cvat-core/src/annotations-objects.js [hunk 1]
```diff
@@ -248,6 +248,19 @@
             this.label = label;
             this.attributes = {};
             this.appendDefaultAttributes(label);
+
+            // Try to keep old attributes if name and inputType matches and old value is still valid
+            for (const attribute of redoLabel.attributes) {
+                for (const oldAttribute of undoLabel.attributes) {
+                    if (
+                        attribute.name === oldAttribute.name
+                        && attribute.inputType === oldAttribute.inputType
```
Comment (bsekachev) [inline cvat-core/src/annotations-objects.js:257]: input types "select" and "radio" are fully compatible
we can also check compatibility additionally between "text" and "radio' or "select"

**Review by bsekachev — COMMENTED** [review]
Hi, sorry for delay in my response.
Generally works fine with shapes and tags, one minor comment about attribute types compatibility. Should we try to keep compatble attributes?

For tracks need to write similar code in Track::_saveLabel at the same file. Do you have an ability to do it? If not, it is okay, we can merge the PR as is. 

Also, please add a line to CONTRIBUTING.md and do ``npm version patch`` in ``cvat-core`` to update its version.

**Comment by timurlenk07** [comment]
I could make the code to work similarly with tracks, though I won't have the possibility to test it.
Will do the line in CONTRIBUTING.md and the version patch, but I did not understand this question of yours:

> Generally works fine with shapes and tags, one minor comment about attribute types compatibility. Should we try to keep compatble attributes?

Could you explain that a bit? Or was it about what you've commented on the code?

**Comment by bsekachev** [comment]
> Generally works fine with shapes and tags, one minor comment about attribute types compatibility. Should we try to keep compatble attributes?
 
Sure, I will try to rephrase.

As you probably know CVAT supports different types of attributes: CHECKBOX, SELECT, RADIO, TEXT and NUMBER.
Internally SELECT and RADIO are the same meaning (one of a list). So, they are compatible from this point of view and if attributes have the same name, we could try to keep it also regardless of different types.

Speaking about TEXT attributes, we definitely can convert SELECT and RADIO to TEXT with no any issues. And we can do a reverse thing if TEXT attribute value is a possible value for RADIO or SELECT attributes.

**Comment by timurlenk07** [comment]
Now I get it. I will try to do these this week.
Should I rebase onto the latest master commit before doing `npm version patch`?

**Comment by timurlenk07** [comment]
Now I only check if the attribute names are the same, and if the old value is still valid in the new attribute. This works even for checkboxes, though I am not sure if that is a desired behaviour.

**Review by bsekachev — APPROVED** [review]

**Comment by bsekachev** [comment]
@dvkruchinin 

Probably we already have a test to change a label, can we update it to cover this case?

**Comment by dvkruchinin** [comment]
> @dvkruchinin
> 
> Probably we already have a test to change a label, can we update it to cover this case?
> Both labels must have an attribute with compatible values.

Sure. I'll check and prepare a corresponding task.

**Pull request merged**